File: Chanseok/프로그래머스/Lv2/17683. ［3차］ 방금그곡/방금그곡.java
class Solution {

    public String solution(String m, String[] musicinfos) {
        String answer = "";

        String heard = changeMusic(m);

        String[] musicinfo;
        String start, end, title, music;
        int time, LongTime = 0, musicLen;
        for(int i = 0, len= musicinfos.length; i < len; ++i) {
            musicinfo = musicinfos[i].split(",");
            start = musicinfo[0];
            end = musicinfo[1];
            title = musicinfo[2];
            music = changeMusic(musicinfo[3]);
            musicLen = music.length();
            
            time = time2min(end) - time2min(start);
            
            if(music.length() > time) music = music.substring(0, time);
            else {
                 
                int repeat = time / musicLen-1;
                int rest = time % musicLen;
                while(repeat-- > 0) {
                    music += music.substring(0, musicLen);
                }
                music += music.substring(0, rest);
            }
            if(music.contains(heard)) {
                if(time > LongTime) {
                    LongTime = time;
                    answer = title;
                }
            }
        }
        if(answer == "") answer = "(None)";
        return answer;
    }

    private boolean isMatch(String heard, String music, int time) {

        int heardLen = heard.length();
        int musicLen = music.length();

        int j, k;
        for (int i = 0; i < musicLen; ++i) {
            k = i;

            for (j = 0; j < heardLen; ++j) {
                if (music.charAt(k) != heard.charAt(j))
                    break;
                k = (++k) % musicLen;
            }

            if(j == heardLen) return true;

        }

        return false;
    }

    private String changeMusic(String m) {
        int delta = 'a' - 'A';
        String n = "";
        char c;
        for(int i = 0, len = m.length(); i < len; ++i) {
            c = m.charAt(i);
            if(i != len-1 && m.charAt(i+1) == '#') {
                n += (char)(delta + c);
                ++i;
            }else {
                n += c;
            }
        }
        return n;
    }


    public int time2min(String time) {
        String[] HHMM = time.split(":");
        return Integer.parseInt(HHMM[0]) * 60 + Integer.parseInt(HHMM[1]);
    }
}

File: Seonguk/프로그래머스/방금그곡/방금그곡.py
def solution(m, musicinfos):
    candi =[]
    dic = {'C#':'H','D#':'I','F#':'J','G#':'K','A#':'L'}
    # #붙은 음 치환하기
    for key, val in dic.items():
        m = m.replace(key,val)
    
    for i,musicinfo in enumerate(musicinfos):
        start,end,name,music = musicinfo.split(',')
        for key, val in dic.items():
            music = music.replace(key, val) # 악보 치환
            
        start, end = list(map(int,start.split(':'))),list(map(int,end.split(':')))
        time = (end[0]-start[0])*60 + end[1]-start[1] # 재생시간을 총 분으로 계산
        
        if time < len(music):
            music = music[:time] # 처음부터 재생된 경우
        else:
            # time만큼 반복 재생
            music = music*(time//len(music)) + music[:(time%len(music))]
        
        # 멜로디가 악보에 있으면 
        if m in music:
            candi.append([name,time,i]) # 후보에 이름, 시간, idx 정보 저장
    
    # 순서, 제목 길이 순으로 정렬
    candi.sort(key=lambda x: (-x[1],x[0]))
    if candi:
        return candi[0][0]
    else:
        return "(None)"
